### PR TITLE
Add ddim inversion pix2pix

### PR DESCRIPTION
**Pull request by patrickvonplaten**
Pix2Pix0: Generate Caption -> Invert -> Generate Image:

```python
import torch
from transformers import BlipForConditionalGeneration, BlipProcessor
from diffusers import DDIMScheduler, DDIMInverseScheduler, StableDiffusionPix2PixZeroPipeline
import requests
from PIL import Image

captioner_id = "Salesforce/blip-image-captioning-base"
processor = BlipProcessor.from_pretrained(captioner_id)
model = BlipForConditionalGeneration.from_pretrained(captioner_id, torch_dtype=torch.float16, low_cpu_mem_usage=True)
sd_model_ckpt = "CompVis/stable-diffusion-v1-4"
pipeline = StableDiffusionPix2PixZeroPipeline.from_pretrained(
    sd_model_ckpt,
    caption_generator=model,
    caption_processor=processor,
    torch_dtype=torch.float16,
    safety_checker=None,
)
pipeline.scheduler = DDIMScheduler.from_config(pipeline.scheduler.config)
pipeline.inverse_scheduler = DDIMInverseScheduler.from_config(pipeline.scheduler.config)
pipeline.enable_model_cpu_offload()

img_url = "https://github.com/pix2pixzero/pix2pix-zero/raw/main/assets/test_images/cats/cat_6.png"
raw_image = Image.open(requests.get(img_url, stream=True).raw).convert("RGB").resize((512, 512))
caption = pipeline.generate_caption(raw_image)

generator = torch.manual_seed(0)
inv_latents = pipeline.invert(caption, image=raw_image, generator=generator).latents

# See the "Generating source and target embeddings" section below to
# automate the generation of these captions with a pre-trained model like Flan-T5.
# automate the generation of these captions with a pre-trained model like Flan-T5 as explained below.
source_prompts = ["a cat sitting on the street", "a cat playing in the field", "a face of a cat"]
target_prompts = ["a dog sitting on the street", "a dog playing in the field", "a face of a dog"]
source_embeds = pipeline.get_embeds(source_prompts, batch_size=2)
target_embeds = pipeline.get_embeds(target_prompts, batch_size=2)
image = pipeline(
    caption,
    source_embeds=source_embeds,
    target_embeds=target_embeds,
    num_inference_steps=50,
    cross_attention_guidance_amount=0.15,
    generator=generator,
    latents=inv_latents,
    negative_prompt=caption,
).images[0]
image.save("edited_image.png")
```

**Source image**:
cat_6

**Generated  image**:
aa (3)

### CODE REVIEW

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py [hunk 1]
```diff
@@ -265,9 +293,36 @@ def enable_sequential_cpu_offload(self, gpu_id=0):
...
+        """
+        if is_accelerate_available() and is_accelerate_version(">=", "0.17.0.dev0"):
+            from accelerate import cpu_offload_with_hook
+        else:
+            raise ImportError("`enable_model_offload` requires `accelerate v0.17.0` or higher.")
+
+        device = torch.device(f"cuda:{gpu_id}")
+
+        hook = None
+        for cpu_offloaded_model in [self.vae, self.text_encoder, self.unet, self.vae]:
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:317]: Repetition in the `self.vae`. You probably meant `self.captioner`?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py [hunk 2]
```diff
@@ -528,15 +568,25 @@ def prepare_latents(self, batch_size, num_channels_latents, height, width, dtype
...
+    def generate_caption(self, images):
         """Generates caption for a given image."""
-        inputs = self._captioner_processor(images=image, return_tensors="pt")
-        outputs = self._captioner.generate(inputs)
-        caption = self._captioner_processor.batch_deocde(outputs, skip_special_tokens=True)[0]
-        if return_image:
-            return caption, inputs["pixel_values"]
-        else:
-            return caption
+        text = "a photography of"
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:574]: Does starting the generation with this cue impact the result?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py [hunk 3]
```diff
@@ -562,6 +612,41 @@ def get_embeds(self, prompt: List[str], batch_size: int = 16) -> torch.FloatTens
 
         return torch.cat(embeds, dim=0).mean(0)[None]
 
+    def prepare_image_latents(self, image, batch_size, dtype, device, generator=None):
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:615]: If it's copied from an existing pipeline let's add the `# Copied from ...` statement. Also, does the image need to be scaled before it passed to this function? If so, is that noted somewhere?

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py [hunk 4]
```diff
@@ -852,7 +915,247 @@ def __call__(
...
+        callback_steps: Optional[int] = 1,
+        cross_attention_kwargs: Optional[Dict[str, Any]] = None,
+        # inversion regularization parameters
+        lambda_ac: float = 20.0,
+        lambda_kl: float = 20.0,
+        num_reg_steps: int = 5,
+        num_ac_rolls: int = 5,
+    ):
+        r"""
+        Function invoked when calling the pipeline for generation.
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:977]: To update.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py (hunk 2 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:574]: It's recommended as shown here: https://huggingface.co/Salesforce/blip-image-captioning-base

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py [hunk 5]
```diff
@@ -852,7 +915,247 @@ def __call__(
...
+                # call the callback, if provided
+                if i == len(timesteps) - 1 or (
+                    (i + 1) > num_warmup_steps and (i + 1) % self.inverse_scheduler.order == 0
+                ):
+                    progress_bar.update()
+                    if callback is not None and i % callback_steps == 0:
+                        callback(i, t, latents)
+
+        x_inv = latents.detach().clone()
+        # reconstruct the image
```
Comment (sayakpaul) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:1151]: (nit): needs to come above the code line and we're reconstructing the noise here and not the image.

File: src/diffusers/schedulers/scheduling_ddim_inverse.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
+# Copyright 2022 Stanford University Team and The HuggingFace Team. All rights reserved.
```
Comment (sayakpaul) [inline src/diffusers/schedulers/scheduling_ddim_inverse.py:1]: Is it from Stanford? The inversion?

File: src/diffusers/schedulers/scheduling_ddim_inverse.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+        t1 = i / num_diffusion_timesteps
+        t2 = (i + 1) / num_diffusion_timesteps
+        betas.append(min(1 - alpha_bar(t2) / alpha_bar(t1), max_beta))
+    return torch.tensor(betas)
+
+
+class DDIMInverseScheduler(SchedulerMixin, ConfigMixin):
+    """
+    Denoising diffusion implicit models is a scheduler that extends the denoising procedure introduced in denoising
+    diffusion probabilistic models (DDPMs) with non-Markovian guidance.
```
Comment (sayakpaul) [inline src/diffusers/schedulers/scheduling_ddim_inverse.py:80]: To be updated.

**Review by sayakpaul — COMMENTED** [review]
LOVE the design 🔥 

Let's make sure we have coverage in totality from the docs.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:316]: That's actually on purpose so that the first self.vae is offloaded when the text encoder is called :sweat_smile:

Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:316]: cc @pcuenca this should work no? Since inversion is img2img

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:614]: yeah this one is a bit different actually

File: src/diffusers/schedulers/scheduling_ddim_inverse.py [hunk 8]
```diff
@@ -0,0 +1,269 @@
+# Copyright 2022 Stanford University Team and The HuggingFace Team. All rights reserved.
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim_inverse.py:1]:
```suggestion
# Copyright 2022 The HuggingFace Team. All rights reserved.
```

**Comment by patrickvonplaten** [comment]
No time for DDIM Inversion tests will add them later: https://github.com/huggingface/diffusers/issues/2399

File: docs/source/en/_toctree.yml [hunk 9]
```diff
@@ -182,6 +182,8 @@
       title: DDIM
     - local: api/schedulers/ddpm
       title: DDPM
+    - local: api/schedulers/ddpm_inverse
+      title: DDPMInverse
```
Comment (patil-suraj) [inline docs/source/en/_toctree.yml:186]:
```suggestion
    - local: api/schedulers/ddim_inverse
      title: DDIMInverse
```

File: docs/source/en/api/schedulers/ddim_inverse.mdx [hunk 10]
```diff
@@ -0,0 +1,21 @@
+<!--Copyright 2022 The HuggingFace Team. All rights reserved.
```
Comment (patil-suraj) [inline docs/source/en/api/schedulers/ddim_inverse.mdx:1]:
```suggestion
<!--Copyright 2023 The HuggingFace Team. All rights reserved.
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py [hunk 11]
```diff
@@ -72,6 +106,65 @@
...
+in the field", "a face of a cat"] >>> target_prompts = ["a dog sitting on the street", "a dog playing in the field", "a
+face of a dog"]
+
+>>> source_embeds = pipeline.get_embeds(source_prompts, batch_size=2) >>> target_embeds =
+pipeline.get_embeds(target_prompts, batch_size=2)
+
+>>> # the latents can then be used to edit a real image >>> image = pipeline( ... caption, ...
+source_embeds=source_embeds, ... target_embeds=target_embeds, ... num_inference_steps=50, ...
+cross_attention_guidance_amount=0.15, ... generator=generator, ... latents=inv_latents, ... negative_prompt=caption,
+... ).images[0] >>> image.save("edited_image.png")
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:143]: should formatted properly.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py [hunk 12]
```diff
@@ -852,7 +997,206 @@ def __call__(
...
+                Whether or not to return a [`~pipelines.stable_diffusion.StableDiffusionPipelineOutput`] instead of a
+                plain tuple.
+            callback (`Callable`, *optional*):
+                A function that will be called every `callback_steps` steps during inference. The function will be
+                called with the following arguments: `callback(step: int, timestep: int, latents: torch.FloatTensor)`.
+            callback_steps (`int`, *optional*, defaults to 1):
+                The frequency at which the `callback` function will be called. If not specified, the callback will be
+                called at every step.
+            lambda_auto_corr (`float`, *optional*, defaults to 20.0):
+                Lambda parameter to control auto correction
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:1074]:
```suggestion
                Lambda parameter to control auto correction.
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py [hunk 13]
```diff
@@ -852,7 +997,206 @@ def __call__(
...
+            lambda_kl (`float`, *optional*, defaults to 20.0):
+                Lambda parameter to control Kullback–Leibler divergence output
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:1076]:
```suggestion
                Lambda parameter to control Kullback–Leibler divergence output.
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py [hunk 14]
```diff
@@ -852,7 +997,206 @@ def __call__(
...
+            num_reg_steps (`int`, *optional*, defaults to 5):
+                Number of regularization loss steps
+            num_auto_corr_rolls (`int`, *optional*, defaults to 5):
+                Number of auto correction roll steps
```
Comment (patil-suraj) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:1080]:
```suggestion
                Number of regularization loss steps.
            num_auto_corr_rolls (`int`, *optional*, defaults to 5):
                Number of auto correction roll steps.
```

File: src/diffusers/schedulers/scheduling_ddim_inverse.py [hunk 15]
```diff
@@ -0,0 +1,269 @@
+# Copyright 2022 The HuggingFace Team. All rights reserved.
```
Comment (patil-suraj) [inline src/diffusers/schedulers/scheduling_ddim_inverse.py:1]:
```suggestion
# Copyright 2023 The HuggingFace Team. All rights reserved.
```

File: src/diffusers/schedulers/scheduling_ddim_inverse.py [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+            prediction type of the scheduler function, one of `epsilon` (predicting the noise of the diffusion
+            process), `sample` (directly predicting the noisy sample`) or `v_prediction` (see section 2.4
+            https://imagen.research.google/video/paper.pdf)
+    """
+
+    order = 1
+
+    @register_to_config
+    def __init__(
+        self,
```
Comment (patil-suraj) [inline src/diffusers/schedulers/scheduling_ddim_inverse.py:117]: This looks good, but to be honest I don't think we should have a new scheduler for this. This is not really a new scheduler, just that the `step` is inverted. What do we think about adding a method called `inverse_step`? 
https://github.com/huggingface/diffusers/issues/2328#issuecomment-1427504951

**Review by patil-suraj — COMMENTED** [review]
Looks great. Thanks a lot for adding this so quickly!  Mostly left some nits. My main comment is: I'm not sure if we need to add a completely new scheduler for just doing the inverse step. Another option would be to add `inverse_step` method to the scheduler.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py (hunk 11 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:143]: ups :sweat_smile:

**Review by patil-suraj — APPROVED** [review]

File: src/diffusers/schedulers/scheduling_ddim_inverse.py (hunk 16 above)
Comment (sayakpaul) [reply src/diffusers/schedulers/scheduling_ddim_inverse.py:117]: So, @patrickvonplaten and I had talked a bit about it yesterday and we both agreed that having it in a seperate scheduler is helpful in terms of a simpler API. 

If we do `inverse _step()` then there is a slight disconnect from the original DDIM paper that didn't have anything for inversion. Since we try to be one with the paper literature, I think it makes sense to have a separate scheduler for this as well.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py (hunk 1 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:371]: Could you elaborate this a bit? Didn't really understand it.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py [hunk 17]
```diff
@@ -852,7 +997,206 @@ def __call__(
...
+
+        # 3. Preprocess image
+        image = preprocess(image)
+
+        # 4. Prepare timesteps
+        self.inverse_scheduler.set_timesteps(num_inference_steps, device=device)
+        timesteps = self.inverse_scheduler.timesteps
+
+        # 5. Prepare latent variables
+        latents = self.prepare_image_latents(image, batch_size, prompt_embeds.dtype, device, generator)
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:1122]: Shouldn't this go before `self._encode_prompt` if we want to take advantage of the cpu offload?

**Review by pcuenca — APPROVED** [review]

File: src/diffusers/schedulers/scheduling_ddim_inverse.py [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+
+    def alpha_bar(time_step):
+        return math.cos((time_step + 0.008) / 1.008 * math.pi / 2) ** 2
+
+    betas = []
+    for i in range(num_diffusion_timesteps):
+        t1 = i / num_diffusion_timesteps
+        t2 = (i + 1) / num_diffusion_timesteps
+        betas.append(min(1 - alpha_bar(t2) / alpha_bar(t1), max_beta))
+    return torch.tensor(betas)
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_ddim_inverse.py:74]:
```suggestion
    return torch.tensor(betas, dtype=torch.float32)
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py (hunk 17 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:1147]: Very good point!

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py (hunk 1 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_pix2pix_zero.py:396]: Hmm, no worries. I will sync up with you offline on this next week.

**Comment by neverix** [comment]
Nice, basically a modern version of #702
